### PR TITLE
Use stateless keyset pagination for searchable import chunks

### PR DESCRIPTION
**Pull request by yocmen**
## What changed

This PR reworks `scout:import` around stateless chunk boundaries and adds a parallel import path for large models.

- Replaced offset/shared-cursor chunking with stateless keyset ranges. Chunk jobs now carry their own primary-key bounds, `CleanLastId` is removed, and `PageScope` is kept only for backwards compatibility.
- `ChunkScope` now filters with the model's qualified key name, so joined/eager-load import queries do not accidentally target an ambiguous or unqualified `id`.
- `DefaultImportSource` now plans chunks from key ranges instead of mutable cache state and guarantees complete coverage without gaps or duplicates across chunks.
- Added faster chunk planning for normal incrementing integer primary keys using `MIN`/`MAX` arithmetic instead of one key-walking query per chunk.
- Added safe fallbacks to seek planning for UUID/string keys, non-incrementing integer keys, and misconfigured models whose declared integer key stores non-numeric values.
- Added `--fast-plan` to plan boundaries from a bare key query, skipping model global scopes/eager-load joins during planning while still applying the full query and `shouldBeSearchable()` when each chunk is fetched.
- Added per-run chunk sizing with `--chunk=N`, plus config-level per-model chunk sizes keyed by `searchableAs()`. Precedence is `--chunk` > `elasticsearch.import.chunk.{searchableAs}` > `elasticsearch.import.chunk.default` > `scout.chunk.searchable`.
- Guarded non-positive chunk sizes so a bad `0`/negative config or option cannot silently build an empty index.
- Made concrete write-index names unique with a random suffix in addition to the timestamp, avoiding same-second collisions between overlapping/stale runs.

## Parallel imports

`scout:import --parallel` now fans chunks out as independent queue jobs coordinated by Redis instead of running every chunk in one import job.

- Added `--parallel` to dispatch a prepare chain (`CleanUp` -> `CreateWriteIndex` -> `DispatchPullChunks`) and then enqueue one `PullChunkJob` per planned chunk.
- Added `--connection` and `--queue` support for the parallel path. Resolution is explicit option first, then `scout.queue`, then the app's default queue connection.
- `--parallel` no longer depends on `scout.queue`; it works with an async app default queue. It fails fast on the `sync` driver unless `--force` is passed.
- Added `--wait` to block after dispatch, show prepare/chunk progress from the Redis run record, and return a final indexed-doc/chunk/elapsed summary.
- `--wait` now distinguishes "no worker picked this up" from "worker started but prepare went silent" by using prepare-stage heartbeats.
- Added `--profile` to log per-chunk fetch/filter/index timings, query counts, payload size, serialization time, bulk time, and lazy-loaded relations for diagnosing slow imports/N+1s.
- Inline sequential imports now print an indexed document count and elapsed time summary. Queued sequential imports still report dispatch only.

## Coordination and safety

- Added `ImportRunStore` and a Redis-backed `RedisImportRunStore` for run status, chunk completion counts, target index name, finalize ownership, and atomic transitions.
- Redis run keys use `{run-token}` hash tags so the Lua scripts are Redis Cluster compatible.
- `--parallel` verifies Redis-backed import coordination is available and rejects the file cache driver for import locks because it cannot provide the required atomic lock behavior.
- Added a renewable per-model `ImportLock` so two imports of the same model cannot race cleanup or alias switching. Different models can still import independently.
- The lock TTL is an inactivity timeout. Sequential imports, prepare stages, chunk jobs, and finalization renew it while work is progressing.
- Same-model duplicate imports are skipped with a warning that includes the exact cache key and a `Cache::forget(...)` recovery hint.
- `CleanUp`, rollback, and final alias switching now re-check the lock owner before deleting or promoting indexes, so a stale run whose lease expired cannot delete or promote over a newer in-progress run.
- Added `FinalizeJob` with a finalize lock so only one chunk winner can refresh the new index, swap aliases, mark success, and release the import lock.
- Added `RollbackImportJob` for failed parallel imports. Failed chunks mark the run failed, remaining chunks skip work, the old index keeps serving, and the unpromoted write index is deleted after a short drain delay.
- Finalization failures get their own `finalize_failed` status so `--wait` can report them separately from chunk failures.

## New configuration and environment variables

New import config lives under `elasticsearch.import`.

- `SCOUT_IMPORT_LOCK_TTL` / `elasticsearch.import.lock_ttl` controls the renewable per-model import lease. Default: `3600`.
- `SCOUT_IMPORT_WAIT_TIMEOUT` / `elasticsearch.import.wait_timeout` controls how long `--parallel --wait` tolerates no observed prepare activity before returning while the queued work continues. Default: `120`.
- `SCOUT_IMPORT_RETRY_TRIES` / `elasticsearch.import.retry.tries` controls total attempts for each parallel chunk job. Default: `1`.
- `SCOUT_IMPORT_RETRY_BACKOFF_BASE` / `elasticsearch.import.retry.backoff_base` sets the base delay, in seconds, for jittered exponential chunk retry backoff. Default: `5`.
- `SCOUT_IMPORT_RETRY_BACKOFF_CAP` / `elasticsearch.import.retry.backoff_cap` caps each chunk retry backoff delay, in seconds. Default: `120`.
- `SCOUT_IMPORT_RETRY_UNTIL` / `elasticsearch.import.retry.retry_until` optionally caps chunk retries by wall-clock seconds. Default: `0` meaning tries-only.
- `SCOUT_IMPORT_ROLLBACK_DELAY` / `elasticsearch.import.rollback_delay` delays failed-run rollback briefly so in-flight chunk jobs can drain before the unpromoted index is deleted. Default: `5`.
- `elasticsearch.import.chunk.default` and `elasticsearch.import.chunk.{searchableAs}` configure default/per-model chunk sizes. These are config values, not env-backed scalar entries.
- `SCOUT_QUEUE_TIMEOUT` remains important operationally: queue `retry_after` or SQS visibility timeout must stay greater than this timeout so a still-running chunk is not made visible again.

Test and local infrastructure now also define Redis envs:

- `REDIS_CLIENT=predis`
- `REDIS_HOST=127.0.0.1` in PHPUnit/GitHub Actions, `redis` in Docker Compose
- `REDIS_PORT=6379`

## Development and CI updates

- Added `predis/predis` as a dev dependency so Redis coordinator tests can run without the PHP Redis extension.
- Added Redis to Docker Compose, including health checks and app container env wiring.
- Added Redis to the GitHub Actions service list and test env.
- Updated `phpunit.xml.dist` with Redis env defaults.
- Updated the local test Make target description to mention Redis.
- Added README coverage for parallel imports, wait behavior, chunk/planning tuning, retry/backoff guidance, Redis coordination, lock semantics, and local package development.
- Added `.DS_Store` to `.gitignore`.
- Tightened type metadata for import sources, config timeout, and index settings/mappings handling.

## Tests added

- Feature coverage for parallel import dispatch, sync-driver rejection, `--force`, async default queue behavior, `--wait` progress/summaries/timeouts, `--chunk`, `--fast-plan`, `--profile`, alias switching, and failed-run behavior.
- Feature coverage for per-model import locks, duplicate-run skipping, lock key messaging, independent imports for different models, lease renewal/release, owner checks, and stale-owner safety.
- Integration coverage for Redis import run store transitions, idempotency, failure/finalization races, finalize ownership, TTL refreshes, and Redis Cluster hash-slot key safety.
- Integration coverage for dispatching chunk jobs, chunk retry/backoff configuration, rollback scheduling, finalize dispatch, zero-record imports, and run-store coordination.
- Integration coverage for keyset chunk coverage, sparse keys, global scope ordering, fast-plan scope stripping, arithmetic planning, seek fallbacks, non-positive chunk safety, and custom/non-incrementing keys.
- Unit coverage for import run store semantics, stage job backoff/retry deadlines/heartbeats/lock renewal, import stage construction, index name uniqueness, and related edge cases.

## Testing status

- `XDEBUG_MODE=off php artisan test` could not run in this package checkout because there is no `artisan` file.
- `XDEBUG_MODE=off ./vendor/bin/phpunit --testdox` starts PHPUnit, but this local run failed during test setup because OpenSearch is not reachable at `127.0.0.1:9200` (`No alive nodes found in your cluster`). Result: 149 tests, 19 assertions, 138 errors, 5 skipped.

### CODE REVIEW

File: src/Searchable/DefaultImportSource.php [hunk 1]
```diff
@@ -49,21 +49,39 @@ public function searchableAs(): string
...
+        // (WHERE key > ? AND key <= ?) rather than OFFSET, so keyset paging
+        // costs the same for the first chunk and the last one. Imports no
+        // longer slow down as they progress through a large table.
+        //
+        // reorder()->orderBy($key) drops any competing ORDER BY (e.g. from a
+        // model global scope or makeAllSearchableUsing) so the key sequence is
+        // strictly monotonic. Without it the boundaries would be sorted by the
+        // wrong column and the id ranges would skip or duplicate rows.
+        $keys = $this->newQuery()->reorder()->orderBy($key)->pluck($key);
+
```
[the remark this answers is not in the document]
Comment (yocmen) [reply src/Searchable/DefaultImportSource.php]: Good catch — this was a memory regression vs the old `count()`-based approach. Fixed in cb7bfed: boundaries are now built with a keyset walk that reads one chunk of keys at a time (`WHERE key > ? ORDER BY key LIMIT chunkSize`), so memory is O(chunk size) instead of O(rows). I used a manual walk rather than lazyById because lazyById's cursor-alias semantics infinite-looped on models with a non-integer/custom primary key (BookWithCustomKey).

File: src/Database/Scopes/PageScope.php [hunk 2]
```diff
@@ -39,19 +38,6 @@ public function __construct(int $page, int $perPage)
...
-            // global scope or makeAllSearchableUsing) would become the primary
-            // sort and cause chunks to skip or duplicate rows. reorder() drops
-            // them all before forPageAfterId sets its own.
-            $builder->reorder();
-            $builder->forPageAfterId($this->perPage, Cache::get('scout_import_last_id', 0), $model->getTable().'.id');
-        } else {
-            $builder->forPage($this->page, $this->perPage);
-        }
+        $builder->forPage($this->page, $this->perPage);
     }
```
[the remark this answers is not in the document]
Comment (yocmen) [reply src/Database/Scopes/PageScope.php:49]: Agreed. Marked `@internal` in cb7bfed with a note pointing to ChunkScope. Note the plain `forPage` behaviour is actually PageScope's original, long-standing semantics — the cache-cursor variant was a recent addition that this PR removes — so this restores prior behaviour rather than changing it. Left the class in place to avoid a breaking removal outside a major release.

File: src/Searchable/DefaultImportSource.php [hunk 3]
```diff
@@ -49,21 +49,57 @@ public function searchableAs(): string
...
-            $chunkSize = (int) config('scout.chunk.searchable', self::DEFAULT_CHUNK_SIZE);
-            $totalChunks = (int) ceil($totalSearchables / $chunkSize);
-
-            return collect(range(1, $totalChunks))->map(function ($page) use ($chunkSize) {
-                $chunkScope = new PageScope($page, $chunkSize);
-
-                return new static($this->className, array_merge($this->scopes, [$chunkScope]));
-            });
-        } else {
+        $chunkSize = (int) config('scout.chunk.searchable', self::DEFAULT_CHUNK_SIZE);
```
[the remark this answers is not in the document]
Comment (yocmen) [reply src/Searchable/DefaultImportSource.php]: Fixed in 8ce3a81 — chunk size is now clamped with `max(1, ...)`, so a misconfigured 0/negative value falls back to 1 instead of producing zero chunks and importing into an empty index. Added a regression test (test_chunked_does_not_silently_skip_rows_when_chunk_size_is_non_positive).
